crates/noirc_abi/src/serialization.rs

```diff
@@ -0,0 +1,233 @@
+            AbiType::Struct { visibility: _, fields } => Self::Struct {
+                fields: fields
+                    .into_iter()
+                    .map(|(name, value)| AbiParameter {
+                        name,
+                        visibility: None,
+                        param_type: Type::from(value),
+                    })
+                    .collect(),
+            },
+        }
+    }
+}
+
+impl AbiType {
+    fn from_type(param_type: Type, visibility: AbiFEType) -> Self {
+        match param_type {
+            Type::Field => AbiType::Field(visibility),
+            Type::Integer { sign, width } => AbiType::Integer { visibility, sign, width },
+            Type::Array { length, typ } => AbiType::Array {
+                visibility,
+                length: length.into(),
+                typ: Box::new(Self::from_type(*typ, visibility)),
+            },
+            Type::Struct { fields } => AbiType::Struct {
+                visibility,
+                fields: fields
+                    .into_iter()
+                    .map(|field| (field.name, AbiType::from_type(field.param_type, visibility)))
+                    .collect(),
+            },
+        }
+    }
+}
+
+impl Serialize for Abi {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        let parameter_vec: Vec<AbiParameter> = self
+            .parameters
+            .clone()
+            .into_iter()
+            .map(|(name, param_type)| {
+                let visibility = Some(param_type.visibility());
+                AbiParameter { name, visibility, param_type: Type::from(param_type) }
+            })
+            .collect();
+
+        parameter_vec.serialize(serializer)
+    }
+}
+
+impl<'de> Deserialize<'de> for Abi {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        let parameters = Vec::<AbiParameter>::deserialize(deserializer)?;
+
+        let parameters = parameters
+            .into_iter()
+            .map(|AbiParameter { name, visibility, param_type }| {
+                (name, AbiType::from_type(param_type, visibility.unwrap()))
+            })
+            .collect();
+
+        Ok(Abi { parameters })
+    }
+
+    fn deserialize_in_place<D>(deserializer: D, place: &mut Self) -> Result<(), D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        // Default implementation just delegates to `deserialize` impl.
+        *place = Deserialize::deserialize(deserializer)?;
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::collections::BTreeMap;
+
+    use crate::{Abi, AbiFEType, AbiType, Sign};
+
+    #[test]
+    fn abi_deserializes_correctly() {
+        let serialized_abi = "
+        [
+            {
+                \"name\": \"thing1\",
+                \"visibility\": \"public\",
+                \"type\": {
+                    \"name\": \"field\"
```

Comment:
I wonder if we can format types a bit differently? Using `{ "name": "field" }` seems odd to me compared to `"field"` but maybe it is worth it since it is more similar to other types? I think just renaming "name" to "kind" would read a bit better.

Reply:
`"kind"` is good. It kinda bugs me how `"name"` refers to both the types and the parameter/field names but couldn't think of a good alternative.